Datepicker

AlexxDk opened the pull request.

AlexxDk committed:
Select the desired date in the calendar

AlexxDk committed:
add 2nd test case: Select the desired date in the calendar

ba-instructor commented on tests/dateSelector.spec.ts:
```diff
@@ -0,0 +1,117 @@
+test.beforeEach(async ({ page }) => {
+    await page.goto("/");
+    await page.getByText(" Owners").click();
+    await page.locator(".dropdown-menu").getByText(" Search").click();
```
Locator can be simplified to just `await page.getByText('Search').click();`

ba-instructor commented on tests/dateSelector.spec.ts:
```diff
@@ -0,0 +1,117 @@
+    await page.locator('#name').fill('Tom')
+    await expect(icon).toHaveClass(/glyphicon-ok/);
+
+    await page.locator('.mat-mdc-button-touch-target').click()
```
Prioritize user-visible locators when possible.
This would be more descriptive: `await page.getByLabel('Open calendar').click();`

ba-instructor commented on tests/dateSelector.spec.ts:
```diff
@@ -0,0 +1,117 @@
+    while (!calendarMonthAndYear?.includes(`${searchMonth} ${searchYear}`)) {
+        await page.locator('[aria-label="Previous month"]').click()
+        calendarMonthAndYear = await page.locator('[aria-label="Choose month and year"]').textContent()
+    }
```
A little bit overkill :)
Looping 10 years in the calendar is quite a lot.
Since the date that should be selected is a constant value and quite far in the past, it would be more reasonable to use a calendar year and month interface to select it
![image](https://github.com/user-attachments/assets/95cd6c28-8668-4476-8863-49891eb68df2)

I ll update the test case to be more descriptive for this step

AlexxDk replied:
Good point)

what about the next approach? 

``` suggestion
    const searchYear = '2014';
    const searchMonth = '05';
    const searchDay = '02';
    const dateToAssert = `${searchYear}/${searchMonth}/${searchDay}`

    await page.getByLabel('Choose month and year').click()
    let calendarHeader = await page.getByLabel('Choose date').textContent()
    let yearFrom = calendarHeader?.split(' – ').map(Number)[0]

    while (yearFrom && (yearFrom > Number(searchYear))) {
        await page.getByLabel('Previous 24 years').click()
        calendarHeader = await page.getByLabel('Choose date').textContent()
        yearFrom = calendarHeader?.split(' – ').map(Number)[0]
    }

    await page.getByText(searchYear).click()
    await page.getByText('May').click()
    await page.locator('[class="mat-calendar-body-cell"]').getByText(searchDay[0] == '0' ? searchDay.slice(1) : searchDay, { exact: true }).click()
    await expect(page.locator('[name="birthDate"]')).toHaveValue(dateToAssert)
```

ba-instructor replied:
Did you write it or ChatGPT? Be honest ))

AlexxDk replied:
Well, to be honest - I did. Exept this small part  `.map(Number)` (only syntax).

About other I think it's obvious was done by myself - because the logic is confusing, and seems that everything can be simplified somehow

ba-instructor replied:
Simplification - is what you always have to start with.
Write the simple code first, then refactor to make it pretty. Starting from the other end using ChatGPT you start digging the rabbit hole that you can't get out of when your code fails and you ll need to fix it.

So I would recommend re-writing this by yourself and making it simple and understandable. Then look at your code again, and see how it can be optimized and simplified even more.

AlexxDk replied:
but it was written by myself) 
ok, will try to rewrite

ba-instructor replied:
Oh... I am sorry.. looks like I didn't get your answer correctly.

AlexxDk replied:
no problem) 
in any case, it seems to need to be simplified

AlexxDk replied:
I did my best 🤞

ba-instructor replied:
This is a great improvement! Looks much cleaner.
But... do you REALLY need that while loop to display the previous 24 years?
The date to select in the test is hardcoded, it will never change according to requirements. So this while loop will always make just one click back. It will never make two clicks because the calendar view has 24 years on the screen. So all you need basically is just this:
```
await page.getByLabel('Choose month and year').click()
await page.getByLabel('Previous 24 years').click()
await page.getByRole('button', { name: searchYear }).click()
await page.getByLabel(`${searchMonth} ${searchYear}`).click()
await page.getByLabel(dateToAssert).click()
```
Super easy to read and easy to understand what this scenario is doing :)

ba-instructor replied:
@AlexxDk this one.

AlexxDk replied:
Now I see, strange.
I thought that once hardcoded date could be changed and in that case no need to update other parts of the code. It is already adapted.
If it is good practice also hardcode steps to achieve 2014 years - no problem. Will do.
Thanks

ba-instructor commented on tests/dateSelector.spec.ts:
```diff
@@ -0,0 +1,117 @@
+
+    const petItem = page.locator('td', { has: page.getByText('Tom') })
+    await expect(petItem).toBeVisible();
+    await expect(petItem).toContainText('2014-05-02')
```
Here is a little test design problem
In line 21,22,23 you have created the constants with the idea in mind to refer to those values later in the test. But here you have used hardcoded values instead.

So either change this assertion to use declared constants, or remove constants and make everywhere just values to assert.
Since the date that should be selected is a strict requirement and is not supposed to be changed in the future, selecting the value right away is also fine.

ba-instructor commented on tests/dateSelector.spec.ts:
```diff
@@ -0,0 +1,117 @@
+    await expect(page.locator('h2')).toHaveText('New Visit')
+    await expect(page.locator('tr', { has: page.getByText('Samantha') })).toContainText('Jean Coleman')
+
+    const calendarInputField = page.locator('[class="mat-mdc-button-touch-target"]')
```
Prioritize user-visible locators when possible.
This would be more descriptive: `page.getByLabel('Open calendar')`

Also the constant name is misleading. It's not the calendar input field. It's more "calendar selector icon" or "date picker icon"

ba-instructor commented on tests/dateSelector.spec.ts:
```diff
@@ -0,0 +1,117 @@
+        }
+        return `${month} ${year}`
+
+    }
```
I see what you was trying to do with this function, but the approach is a little bit confusing, here is why...
1. Why the default return format for the function is MM-YYYY? If I use function with a name "formatDate" I expect from it some kind of consistency, like returning different combinations of MM-DD-YYYY or DD-MM-YYYY and so on. But why MM-YYYY? I understand that this is because what is needed in the script. But from the design on the function perspective - it is confusing

2. Why the "calendar" is the default value for the page variable? This value is not used anywhere, so it can be XYZ and it would work the same.

3. Expected variables like "visit" or "owner" I assume come from the name of the page where this calendar is used. But from the function design perspective it's incorrect. The function should not technically care about pages. This function is about dates and it can format the date no matter on which page it was used. It can format based on the requested format. So it makes more sense to make an expected input as a format type that this function should perform.

So something like this, would make the function more meaningful:
```
function formatDate(date: Date, expectedFormat: string) {
    if (!expectedFormat) {
        throw new Error('The expectedFormat parameter is required.');
    }

    const year = date.getFullYear();
    const month = date.toLocaleString('EN-US', { month: '2-digit' });
    const day = date.toLocaleString('EN-US', { day: '2-digit' });

    if (expectedFormat === 'YYYY/MM/DD') {
        return `${year}/${month}/${day}`;
    } else if (expectedFormat === 'YYYY-MM-DD') {
        return `${year}-${month}-${day}`;
    } else if (expectedFormat === 'MM YYYY') {
        return `${month} ${year}`;
    } else {
        throw new Error('The provided expectedFormat is not supported.');
    }
}
```
Also in the test when you call the function it becomes more readable with understandable expectations of what function should return:
`const formattedDateOnTheNewVisitPage = formatDate(date, 'YYYY/MM/DD');`

Also I was using here `toLocaleString()` method as a more optimal way to format the date (how it was shown in the class)

AlexxDk replied:
got it,
thanks

ba-instructor commented on tests/dateSelector.spec.ts:
```diff
@@ -0,0 +1,117 @@
+test('Select the desired date in the calendar 2', async ({ page }) => {
+    await page.getByRole('link', { name: 'Jean Coleman' }).click()
+
+    const petItem = page.locator('td', { has: page.getByText('Samantha') })
```
This constant was used just once so not needed

ba-instructor commented on tests/dateSelector.spec.ts:
```diff
@@ -0,0 +1,117 @@
+    const firstDateValue = await firstsRowElement.locator('td').first().textContent()
+    const secondDateValue = await secondRowElement.locator('td').first().textContent()
+
+    if (firstDateValue && secondDateValue) {
```
Do you expect that any of those variables can be false?

AlexxDk replied:
I added this checking to handle Typescript error: 
<img width="978" alt="Знімок екрана 2024-07-23 о 00 29 44" src="https://github.com/user-attachments/assets/fc3367bb-21b4-4020-ab04-6d00b26b2ca1">

ba-instructor replied:
I see, makes sense.
Add the exclamation sign at the end of the argument. That way you tell TypeScript that you are confident that the value will not be null or undefined:
```
const firstDate = new Date(firstDateValue!);
const secondDate = new Date(secondDateValue!);
expect(firstDate > secondDate).toBeTruthy();
```

AlexxDk replied:
ok, thanks

ba-instructor commented on tests/dateSelector.spec.ts:
```diff
@@ -0,0 +1,117 @@
+    const addVisitBtnOnTheVisitPage = page.getByRole('button', { name: 'Add Visit' })
+    await addVisitBtnOnTheVisitPage.click()
+
+    const newVisitInfo = await page.locator('app-pet-list', { has: page.getByText('Samantha') }).locator('app-visit-list tr td').first()
```
`await` is not needed here.
What is newVisitInfo? By reading the code it's not clear what this locator is referring to
Did you mean "firstVisitRowInfoForSamantha"?

ba-instructor reviewed (commented):
Please check the comments below.
Overall pretty good. Don't forget about meaningful naming for constants and keep the integrity of test design.

AlexxDk committed:
fixes after review

AlexxDk committed:
simplify date selection logic

ba-instructor commented:
@AlexxDk I see you have requested a review, but I don't see new commits.

AlexxDk commented:
> @AlexxDk I see you have requested a review, but I don't see new commits.

<img width="423" alt="Знімок екрана 2024-07-25 о 21 12 14" src="https://github.com/user-attachments/assets/ee9374c1-ab0a-4a8b-9dce-7edff63d52e3">


@ba-instructor  https://github.com/AlexxDk/playwright-homework/pull/6/commits/0ba288c7954a41e7b6b7b3741a60b9eabd46e048 last changes, simplified

ba-instructor commented:
> @ba-instructor [0ba288c](https://github.com/AlexxDk/playwright-homework/commit/0ba288c7954a41e7b6b7b3741a60b9eabd46e048) last changes, simplified

After that I have added one comment. Please check above.

AlexxDk commented:
> > @ba-instructor [0ba288c](https://github.com/AlexxDk/playwright-homework/commit/0ba288c7954a41e7b6b7b3741a60b9eabd46e048) last changes, simplified
> 
> After that I have added one comment. Please check above.

Unfortunately I don’t see. Which one ?

ba-instructor reviewed (commented):
I think I didn't submit the comment. Now it should be visible.

AlexxDk committed:
remove whileloop for selecting needed date

ba-instructor reviewed (approved):
All right, looks good! )
